staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -0,0 +1,229 @@
+	oldLease, oldLeaseOk := isValidIdentityLease(oldObj, identityLeaseLabelSelector)
+	newLease, newLeaseOk := isValidIdentityLease(newObj, identityLeaseLabelSelector)
+
+	switch {
+	case !oldLeaseOk && !newLeaseOk:
+		return
+	case !oldLeaseOk && newLeaseOk:
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer to discovery cache", "serverID", newLease.Name)
+		}
+	case oldLeaseOk && !newLeaseOk:
+		h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+			delete(cache, oldLease.Name)
+		})
+	case oldLeaseOk && newLeaseOk:
+		// Delete old discovery info if holderIdentity changed, implying
+		// the sever was restarted.
+		if oldLease.Name == newLease.Name && *oldLease.Spec.HolderIdentity != *newLease.Spec.HolderIdentity {
+			h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+				delete(cache, oldLease.Name)
+			})
+		}
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer to discovery cache", "serverID", newLease.Name)
+		}
+	}
+}
+
+func (h *peerProxyHandler) addPeerDiscoveryInfoToCache(serverIdentityLease *v1.Lease) error {
+	// Don't record our own discovery info.
+	if serverIdentityLease.Name == h.serverId {
+		return nil
+	}
+
+	var discoveryDoc *peerAggDiscoveryInfo
+	var ok bool
+	h.withPeerAggDiscoveryCacheRLock(func(cache map[string]*peerAggDiscoveryInfo) {
+		discoveryDoc, ok = cache[serverIdentityLease.Name]
+	})
+
+	if !ok ||
+		len(discoveryDoc.servedResources) == 0 ||
+		discoveryDoc.holderIdentity != *serverIdentityLease.Spec.HolderIdentity {
+		return h.fetchNewDiscoveryFor(serverIdentityLease.Name, *serverIdentityLease.Spec.HolderIdentity)
+	}
+
+	return nil
+}
+func isValidIdentityLease(obj interface{}, identityLeaseLabelSelector string) (*v1.Lease, bool) {
+	lease, ok := obj.(*v1.Lease)
+	if !ok {
+		klog.Error(fmt.Errorf("invalid lease object provided, received type: %T", obj))
+		return nil, false
+	}
+
+	parts := strings.Split(identityLeaseLabelSelector, "=")
+	if len(parts) != 2 {
+		klog.ErrorS(nil, "invalid identityLeaseLabelSelector provided, must be of the form key=value", "Received", identityLeaseLabelSelector)
+		return nil, false
+	}
+
+	identityLeaseComponentLabelKey := parts[0]
+	identityLeaseComponentLabelValue := parts[1]
+	if lease == nil || lease.Labels == nil ||
+		lease.Labels[identityLeaseComponentLabelKey] != identityLeaseComponentLabelValue {
+		klog.Error(fmt.Errorf("invalid lease object provided, must be an apiserver identity lease"))
+		return nil, false
+	}
+
+	if lease.Spec.HolderIdentity == nil {
+		klog.Error(fmt.Errorf("invalid lease object provided, missing holderIdentity in lease obj"))
+		return nil, false
+	}
+
+	return lease, true
+}
+
+func (h *peerProxyHandler) fetchNewDiscoveryFor(serverID string, holderIdentity string) error {
```

Comment:
If this fails (let's say due to a temporary network issue), when's the next time we'll issue the request again?

Reply:
From my observation in the integration test, when a lease is renewed, it shows up as an Add event in the informer.  Therefore addPeerDiscoveryInfoToCache() is called everytime a lease is renewed and we fetch new discovery only if one of the following is true, as a part of addPeerDiscoveryInfoToCache():

- peerDiscoveryInfo for the given serverID is not found in cache   
- peerDiscoveryInfo is found in the cache but has 0 servedResources specified (may happen during network errors for prev. discovery call)
- peerDiscoveryInfo is found in the cache but holderIdentity has changed

Reply:
ok, that's good. How about partial successes like only a success on `/api` and `/apis` fail? cache will have an incomplete list of servedResources but since it's greater than 0, it'll skip right?

Reply:
Resolving, since now we have a sync loop that will retry on errors returned from discovery.